connector/topologyconnector/internal/signal_traverser.go

```diff
@@ -0,0 +1,272 @@
+package internal
+
+import (
+	"context"
+	"sync"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+)
+
+// ------------------------
+// Metric traversal
+// ------------------------
+
+// MetricsTraverser provides a visitor-based traversal over OpenTelemetry metric data.
+// It caches attribute maps at various levels to avoid repeated conversions and reduce
+// allocation overhead during traversal.
+//
+// Fields:
+//
+//	metrics           - the raw pmetric.Metrics to traverse
+//	resourceAttrCache  - caches resource-level attributes by resourceIndex
+//	scopeAttrCache     - caches scope-level attributes by [resourceIndex, scopeIndex]
+//	metricMetaCache    - caches metric-level metadata by [resourceIndex, scopeIndex, metricIndex]
+//	dpAttrCache        - caches datapoint-level attributes by [resourceIndex, scopeIndex, metricIndex, datapointIndex]
+type MetricsTraverser struct {
+	metrics pmetric.Metrics
+
+	resourceAttrCache sync.Map // map[int]map[string]any
+	scopeAttrCache    sync.Map // map[[2]int]map[string]any
+	metricMetaCache   sync.Map // map[[3]int]map[string]any
+	dpAttrCache       sync.Map // map[[4]int]map[string]any
+}
+
+type SignalTraverser interface {
+	Traverse(ctx context.Context, mappingVisitor MappingVisitor)
+}
+
+func NewMetricsTraverser(metrics pmetric.Metrics) *MetricsTraverser {
+	return &MetricsTraverser{
+		metrics: metrics,
+	}
+}
+
+func (t *MetricsTraverser) resourceAttrs(i int, rm pmetric.ResourceMetricsSlice) map[string]any {
+	if v, ok := t.resourceAttrCache.Load(i); ok {
+		if cached, typeOk := v.(map[string]any); typeOk {
+			return cached
+		}
+	}
+	m := rm.At(i).Resource().Attributes().AsRaw()
+	t.resourceAttrCache.Store(i, m)
+	return m
+}
+
+func (t *MetricsTraverser) scopeAttrs(i, j int, sm pmetric.ScopeMetricsSlice) map[string]any {
+	key := [2]int{i, j}
+	if v, ok := t.scopeAttrCache.Load(key); ok {
+		if cached, typeOk := v.(map[string]any); typeOk {
+			return cached
+		}
+	}
+	m := sm.At(j).Scope().Attributes().AsRaw()
+	t.scopeAttrCache.Store(key, m)
+	return m
+}
+
+func (t *MetricsTraverser) metricAttrs(i, j, k int, m pmetric.Metric) map[string]any {
```

Comment:
These parameter names are very cryptic, can we rename them such they represent something (I'm guessing `resource`, `scope`, `metric`, `datapoint` indexes)